File: apps/android/contract/src/main/kotlin/ai/deepseek/dsh/contract/RemoteFailureClasses.kt
package ai.deepseek.dsh.contract

/**
 * Codes with agreed cross-Client semantics, mirrored from
 * REMOTE_FAILURE_CLASSES in @deepseek-ai/dsh-typert-protocol. The vocabulary
 * is merge-extensible: codes without an entry deliberately resolve to
 * [RemoteFailureClass.UNKNOWN] and must stay presentable as opaque
 * diagnostics. The repository test asserts this mirror equals the generated
 * projection of the TypeScript authority and references only codes the
 * published envelope schema declares.
 */
object RemoteFailureClasses {

    /** Mirror of the TypeScript authority; kept exact by EnvelopeSchemaTest. */
    val BY_CODE: Map<String, RemoteFailureClass> = mapOf(
        "gateway/authentication-required" to RemoteFailureClass.AUTHENTICATION,
        "gateway/permission-denied" to RemoteFailureClass.PERMISSION,
        "subagent/unauthorized" to RemoteFailureClass.PERMISSION,
        "gateway/host-not-ready" to RemoteFailureClass.HOST_STATE,
        "gateway/protocol-unsupported" to RemoteFailureClass.COMPATIBILITY,
        "host/protocol-unsupported" to RemoteFailureClass.COMPATIBILITY,
        "host/capability-unavailable" to RemoteFailureClass.COMPATIBILITY,
        "host/description-invalid" to RemoteFailureClass.CARRIER_INVALID,
        "gateway/preparation-unavailable" to RemoteFailureClass.CARRIER_INVALID,
        "gateway/stream-invalid" to RemoteFailureClass.CARRIER_INVALID,
        "gateway/transport-interrupted" to RemoteFailureClass.TRANSPORT,
        "gateway/connection-unavailable" to RemoteFailureClass.TRANSPORT,
        "revision-conflict" to RemoteFailureClass.CONFLICT,
        "session/revision-conflict" to RemoteFailureClass.CONFLICT,
        "session/not-found" to RemoteFailureClass.UNAVAILABLE,
        "session/queue-item-not-found" to RemoteFailureClass.UNAVAILABLE,
        "subagent/not-found" to RemoteFailureClass.UNAVAILABLE,
        "agent-preset/not-found" to RemoteFailureClass.UNAVAILABLE,
        "workspace/not-found" to RemoteFailureClass.UNAVAILABLE,
        "workspace-file/not-found" to RemoteFailureClass.UNAVAILABLE,
        "presented-file/not-found" to RemoteFailureClass.UNAVAILABLE,
    )

    /**
     * Classify one Remote failure code for presentation.
     * @param code code carried by a Remote failure envelope, known or unknown.
     * @returns the agreed class; UNKNOWN for codes without shared semantics.
     */
    fun classify(code: String): RemoteFailureClass = BY_CODE[code] ?: RemoteFailureClass.UNKNOWN
}

File: apps/apple/contract/Sources/DSHContract/RemoteFailureClasses.swift
/// Codes with agreed cross-Client semantics, mirrored from
/// REMOTE_FAILURE_CLASSES in @deepseek-ai/dsh-typert-protocol. The vocabulary
/// is merge-extensible: unlisted codes deliberately resolve to `unknown` and
/// must stay presentable as opaque diagnostics. Tests assert this mirror
/// equals the generated projection of the TypeScript authority.
public enum RemoteFailureClasses {
    /// Mirror of the TypeScript authority; kept exact by the package tests.
    public static let byCode: [String: RemoteFailureClass] = [
        "gateway/authentication-required": .authentication,
        "device/admission-expired": .authentication,
        "device/replay-detected": .authentication,
        "gateway/permission-denied": .permission,
        "subagent/unauthorized": .permission,
        "gateway/host-not-ready": .hostState,
        "gateway/protocol-unsupported": .compatibility,
        "host/protocol-unsupported": .compatibility,
        "host/capability-unavailable": .compatibility,
        "host/description-invalid": .carrierInvalid,
        "gateway/preparation-unavailable": .carrierInvalid,
        "gateway/stream-invalid": .carrierInvalid,
        "gateway/transport-interrupted": .transport,
        "gateway/connection-unavailable": .transport,
        "revision-conflict": .conflict,
        "session/revision-conflict": .conflict,
        "session/not-found": .unavailable,
        "session/queue-item-not-found": .unavailable,
        "subagent/not-found": .unavailable,
        "agent-preset/not-found": .unavailable,
        "workspace/not-found": .unavailable,
        "workspace-file/not-found": .unavailable,
        "workspace-file/not-regular-file": .unavailable,
        "workspace-file/not-directory": .unavailable,
        "presented-file/not-found": .unavailable,
        "session/attachment-invalid": .invalidInput,
        "subagent/attachment-invalid": .invalidInput,
        "session/title-invalid": .invalidInput,
        "agent-preset/invalid": .invalidInput,
        "workspace/invalid-path": .invalidInput,
    ]

    /// Classify one Remote failure code for presentation.
    /// - Parameter code: code carried by a Remote failure envelope.
    /// - Returns: the agreed class; `unknown` for codes without shared semantics.
    public static func classify(_ code: String) -> RemoteFailureClass {
        byCode[code] ?? .unknown
    }
}
